### ISSUE TITLE
Internal coalton bug: Invalid state reached in rewrite-pattern-vars

### ISSUE THREAD

**Issue by Inc0n** [coalton-lang/coalton#767]
I am using commit 3dcf552, which is head right now.
SBCL 2.2.9. 
Mac M1.

> If you are seeing this, please file an issue on Github.
   [Condition of type COALTON-IMPL/UTIL:COALTON-BUG]
Code:
```
(coalton-toplevel
  (define-type (Dequeue :a)
	(Dequeue (List :a) (List :a))		; Dequeue front back
	)
  (declare push (:a -> (Dequeue :a)))
  (define (push x q)
	(match q
	  ((Dequeue '() '())
	   (let ((y (list x)))
		 (Dequeue y y)))
	  ((Dequeue f b)
	   (Dequeue (cons x f) b)))))
```

Backtrace:
```
  0: (COALTON-IMPL/UTIL:COALTON-BUG "Invalid state reached in rewrite-pattern-vars")
  1: (COALTON-IMPL/AST/PATTERN:REWRITE-PATTERN-VARS #S(COALTON-IMPL/AST/PATTERN:PATTERN-VAR :ID NIL-158) #S(COALTON-IMPL/ALGORITHM/IMMUTABLE-MAP:IMMUTABLE-MAP :DATA #{| (NIL-157 COMMON-LISP:NIL) (NIL-158 C..
  2: (COALTON-IMPL/AST/PATTERN:REWRITE-PATTERN-VARS #S(COALTON-IMPL/AST/PATTERN:PATTERN-CONSTRUCTOR :NAME COMMON-LISP:QUOTE :PATTERNS (#S(COALTON-IMPL/AST/PATTERN:PATTERN-VAR :ID NIL-158))) #S(COALTON-IMPL..
  3: (COALTON-IMPL/AST/PATTERN:REWRITE-PATTERN-VARS #S(COALTON-IMPL/AST/PATTERN:PATTERN-CONSTRUCTOR :NAME DEQUEUE :PATTERNS (#S(COALTON-IMPL/AST/PATTERN:PATTERN-CONSTRUCTOR :NAME COMMON-LISP:QUOTE :PATTERN..
  4: (COALTON-IMPL/TYPECHECKER/DERIVE-TYPE::DERIVE-MATCH-BRANCHES-TYPE #S(COALTON-IMPL/TYPECHECKER/TYPES:TAPP :FROM #S(COALTON-IMPL/TYPECHECKER/TYPES:TAPP :FROM #S(COALTON-IMPL/TYPECHECKER/TYPES:TYCON :NAM..
  5: ((:METHOD COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-EXPRESSION-TYPE (COALTON-IMPL/AST/NODE:NODE-MATCH COMMON-LISP:T COMMON-LISP:T)) #S(COALTON-IMPL/AST/NODE:NODE-MATCH :UNPARSED (MATCH Q #1=(# #) #2..
  6: ((:METHOD COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-EXPRESSION-TYPE :AROUND (COALTON-IMPL/AST/NODE:NODE COMMON-LISP:T COMMON-LISP:T)) #S(COALTON-IMPL/AST/NODE:NODE-MATCH :UNPARSED (MATCH Q #1=(# #) ..
  7: ((:METHOD COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-EXPRESSION-TYPE (COALTON-IMPL/AST/NODE:NODE-ABSTRACTION COMMON-LISP:T COMMON-LISP:T)) #S(COALTON-IMPL/AST/NODE:NODE-ABSTRACTION :UNPARSED (FN (X Q..
  8: ((:METHOD COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-EXPRESSION-TYPE :AROUND (COALTON-IMPL/AST/NODE:NODE COMMON-LISP:T COMMON-LISP:T)) #S(COALTON-IMPL/AST/NODE:NODE-ABSTRACTION :UNPARSED (FN (X Q) #1..
  9: (COALTON-IMPL/TYPECHECKER/DERIVE-TYPE::DERIVE-BINDING-TYPE PUSH #S(COALTON-IMPL/TYPECHECKER/TYPES:TAPP :FROM #S(COALTON-IMPL/TYPECHECKER/TYPES:TAPP :FROM #S(COALTON-IMPL/TYPECHECKER/TYPES:TYCON :NAME ..
 10: (COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-EXPL-TYPE (PUSH . #S(COALTON-IMPL/AST/NODE:NODE-ABSTRACTION :UNPARSED (FN # #1=#) :VARS (X-155 Q-156) :SUBEXPR #S(COALTON-IMPL/AST/NODE:NODE-MATCH :UNPARSE..
 11: (COALTON-IMPL/TYPECHECKER/DERIVE-TYPE:DERIVE-BINDINGS-TYPE COMMON-LISP:NIL ((PUSH . #S(COALTON-IMPL/AST/NODE:NODE-ABSTRACTION :UNPARSED # :VARS # :SUBEXPR # :NAME-MAP #))) #<COMMON-LISP:HASH-TABLE :TE..
 12: (COALTON-IMPL::PROCESS-TOPLEVEL-VALUE-DEFINITIONS ((DEFINE (PUSH X Q) (MATCH Q # #))) #<COMMON-LISP:HASH-TABLE :TEST COMMON-LISP:EQL :COUNT 1 {7008ADC573}> #<COMMON-LISP:PACKAGE "LISP-HDL.COALTON"> #<..
 13: (COALTON-IMPL::PROCESS-COALTON-TOPLEVEL ((DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))) #<COMMON-LISP:PACKAGE "LISP-HDL.COALTON"> #<COALTON-IMPL/T..
 14: ((COMMON-LISP:MACRO-FUNCTION COALTON-TOPLEVEL) (COALTON-TOPLEVEL (DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))) #<unused argument>)
 15: (COMMON-LISP:MACROEXPAND-1 (COALTON-TOPLEVEL (DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))) #<NULL-LEXENV>)
 16: (COMMON-LISP:MACROEXPAND (COALTON-TOPLEVEL (DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))) #<NULL-LEXENV>)
 17: (SB-INT:SIMPLE-EVAL-IN-LEXENV (COALTON-TOPLEVEL (DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))) #<NULL-LEXENV>)
 18: (COMMON-LISP:EVAL (COALTON-TOPLEVEL (DEFINE-TYPE (DEQUEUE :A) (DEQUEUE # #)) (DECLARE PUSH (:A -> #)) (DEFINE (PUSH X Q) (MATCH Q # #))))
 19: ((COMMON-LISP:LAMBDA COMMON-LISP:NIL :IN SWANK:INTERACTIVE-EVAL))
```

**Comment by Inc0n** [comment]
However, I am hitting a different error now, after applying the fix mentioned in the previous comment.
```
The function COALTON-IMPL::SEXP-FMT is undefined.
```
I am fairly confused on what to do with that.

**Comment by eliaslfox** [comment]
The first issue should be fixed by #770. The second one should be fixed by #769. 

My guess is that that the compiler was trying to call sexp-fmt to warn about pattern matching.  This defines a variable called None and matches on anything:
```
(define (f x)
  (None x))
```

This matches on the constructor `None`.
```
(define (f x)
  ((None) x))
```

**Comment by Inc0n** [comment]
It works great now, thanks. I just need to fix my other logic errors...
